README
======================================================================

		Zentyal Firewall module by MAC

======================================================================


This module allow [internal] traffic based on MAC for every object member.



1.- Create a Objects and add members with MAC, hostname and IP
    (use 127.1.1.x for members without static address)

2.- Enable module

3.- Watch that you can use internal network if your MAC is registered

4.- zentyal-dhcp package use ISC dhcp3-server that CAN'T be firewalled
    Uninstall it and install dnsmasq for block DHCP traffic.


debian/changelog
zentyal-fwmac (5.0.0) xenial; urgency=medium

  * Change modInstance 'objects' => 'network'

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Sat, 25 Nov 2017 23:30:05 +0100

zentyal-fwmac (3.5.1) trusty; urgency=medium

  * Rewrite for Zentyal 3.5

 -- Mario Izquierdo (mariodebian) <mariodebian@gmail.com>  Thu, 17 Jul 2014 11:10:15 +0200

zentyal-fwmac (2.2.1) lucid; urgency=low

  * Initial version

 -- Mario Izquierdo Rodríguez (Thinetic Systems SLNE) <mario.izquierdo@thinetic.es>  Tue, 10 Jan 2012 15:50:24 +0100


usr/share/perl5/EBox/FwMACFirewall.pm
# Copyright (C) 2011 Mario Izquierdo (mariodebian) for Comunidad de Madrid
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

package EBox::FwMACFirewall;
use strict;
use warnings;

use base 'EBox::FirewallHelper';


use EBox::Global;
use EBox::Config;
use EBox::Firewall;
use EBox::Gettext;
use EBox::Sudo;

sub new
{
        my $class = shift;
        my %opts = @_;
        my $self = $class->SUPER::new(@_);
        bless($self, $class);
        return $self;
}

sub input
{
	my $self = shift;
	my @rules = ();

	my $net = EBox::Global->modInstance('network');
	my @ifaces = @{$net->InternalIfaces()};
	my $obj = EBox::Global->modInstance('network');

	EBox::Sudo::root('/sbin/iptables -N allowmacs >/dev/null 2>&1 || true');

	foreach my $ifc (@ifaces) {
		# add at first position of INPUT chain
		EBox::Sudo::root("/sbin/iptables -nvL INPUT| egrep -q '.*(allowmacs).*($ifc).*' || /sbin/iptables -I INPUT -i $ifc -j allowmacs");
	}

	foreach my $object (@{$obj->objects}) {
		my $obj_desc = $obj->objectDescription($object->{id});

		# Only OBJECT name => allowed
		#next unless ($obj_desc eq 'allowmacs');

		my $members = $obj->objectMembers($object->{id});
		foreach my $member (@{$members}) {
			my $mac = $member->{macaddr};

			defined($mac) or next;
			($mac ne "") or next;

			my $address = $member->{ipaddr};
			my $name = $member->{name};
			#print "IP=$address MAC=$mac NAME=$name\n";
			foreach my $ifc (@ifaces) {
				# FIXME ACCEPT or RETURN
				my $r="-i $ifc -m mac --mac-source $mac -m comment --comment 'IP=$address NAME=$name OBJ=$obj_desc' -j ACCEPT";
				#push(@rules, $r);
				push(@rules, { 'rule' => $r, 'chain' => 'allowmacs' });
			}
		}

	}

	# reject if any rule
	my $size=scalar @rules;
	if ( $size > 0 ) {
		# reject other traffic
		foreach my $ifc (@ifaces) {
			#my $r2="-i $ifc -j LOG --log-prefix '[NO ALLOW MAC $ifc] '";
			#push(@rules, { 'rule' => $r2, 'chain' => 'allowmacs' });

			#my $r="-i $ifc -j DROP";
			my $r="-i $ifc -j idrop";
			push(@rules, { 'rule' => $r, 'chain' => 'allowmacs' });
		}
	}

	return \@rules;
}



1;


usr/share/perl5/EBox/FwMAC.pm
# Copyright (C) 2011 Mario Izquierdo (mariodebian) for Comunidad de Madrid
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

package EBox::FwMAC;

use strict;
use warnings;

use base qw(EBox::Module::Service
            EBox::FirewallObserver
            );

use EBox::Exceptions::DataExists;
use EBox::Gettext;
use EBox::Menu::Item;
use EBox::Service;
use EBox::Sudo qw ( :all );
use EBox::Validate qw ( :all );
use File::Temp qw( tempfile tempdir );
use EBox::FwMACFirewall;

use EBox::Dashboard::Section;
use EBox::Dashboard::List;

sub _create
{
    my $class = shift;
    my $self = $class->SUPER::_create(name => 'fwmac',
                      printableName => 'Cortafuegos por MAC',
                      @_);
    bless($self, $class);
    return $self;
}

# Method: menu
#
#       Overrides EBox::Module method.
sub menu
{
    my ($self, $root) = @_;
    $root->add(new EBox::Menu::Item('url' => 'FwMAC/Index',
                                    'text' => $self->printableName(),
                                    'separator' => 'Gateway',
                                    'icon' => 'firewall',
                                    'order' => 229));
}

sub _generateDNSmasq
{
    my $log = EBox::logger;
    $log->info("FWMAC => _generateDNSmasq()");

    my $dhcpFile = new File::Temp(TEMPLATE => 'dnsmasq.mac.conf-XXXXXX',
                                 DIR      => EBox::Config::tmp());

    open DHCPFILE, ">$dhcpFile";
    my $obj = EBox::Global->modInstance('network');

    foreach my $object (@{$obj->objects}) {
        my $obj_desc = $obj->objectDescription($object->{id});
        
        my $members = $obj->objectMembers($object->{id});
        foreach my $member (@{$members}) {
            my $mac = $member->{macaddr};
            
            defined($mac) or next;
            ($mac ne "") or next;

            my $address = $member->{ipaddr};
            $address =~ s/\/32//;
            my $name = $member->{name};
            
            $log->info("FWMAC => firewallHelper dhcp-host=$mac,$address,$name");
            if ( index($address, '127.1') == -1 ) {
                print DHCPFILE "dhcp-host=$mac,$address,$name\n";
            }
        }
    }

    close (DHCPFILE);
    # /etc/init.d/dnsmasq stop
    # :> /var/lib/misc/dnsmasq.leases
    # sleep 1
    # /etc/init.d/dnsmasq start

    EBox::Sudo::root("cp $dhcpFile /etc/dnsmasq.mac.conf",
                     "service dnsmasq stop",
                     "cat /dev/null > /var/lib/misc/dnsmasq.leases",
                     "sleep 1",
                     "service dnsmasq start");
}

sub _preServiceHook
{
    my ($self, $enabled) = @_;
    
    my $log = EBox::logger;
    $log->info("FWMAC => _preServiceHook");
    
    if(! $enabled) {
        EBox::Sudo::root('for n in $(iptables -nvL INPUT --line-numbers | grep allowmacs | tac | awk \'{print $1}\'); do iptables -D INPUT $n ; done',
                         "iptables -F allowmacs >/dev/null 2>&1||true",
                         "iptables -X allowmacs >/dev/null 2>&1||true");
        
    }



    return $self->SUPER::_preServiceHook($enabled);
}

sub _postServiceHook
{
    my ($self, $enabled) = @_;
    $self->_generateDNSmasq();
    return $self->SUPER::_postServiceHook($enabled);
}

sub firewallHelper
{
    my ($self) = @_;
    
    if ($self->isEnabled()){
        return new EBox::FwMACFirewall();
    }
    return undef;
}


sub dnsmasqLeasesWidget
{
    my ($self, $widget) = @_;

    my $section = new EBox::Dashboard::Section('dnsmasqleases');
    $widget->add($section);
    #my $titles = [__('IP address'),__('MAC address'), __('Host name'), __('Date')];
    my $titles = [__('IP address'),__('MAC address'), __('Host name')];

    my $rows = {};

    open(my $FD, '/var/lib/misc/dnsmasq.leases');
    while(my $line = <$FD>) {
        chomp($line);
        my ($datetime, $mac, $ip, $name) = split(' ',$line);
        my ($ipa, $ipb, $ipc, $ipd) = split('\.', $ip);

        #my ($S,$M,$H,$d,$m,$Y) = localtime($datetime,);
        #$m += 1;
        #$Y += 1900;
        #my $dt = sprintf("%04d-%02d-%02d %02d:%02d:%02d", $Y,$m,$d, $H,$M,$S);

        #$rows->{$ipd} = [$ip, $mac, $name, $dt];
        $rows->{"dnsmasq-$ipa-$ipb-$ipc-$ipd"} = [$ip, $mac, $name];
    }
    close($FD);

    # sort ids array
    my $sorted=[];
    #foreach my $id (sort {$a<=>$b} keys (%{$rows})) {
    foreach my $id (sort {lc $a cmp lc $b} keys (%{$rows})) {
      #print "id=$id\n";
      push(@{$sorted}, $id);
    }

    $section->add(new EBox::Dashboard::List(undef, $titles, $sorted, $rows));
}

sub widgets
{
    return {
        'dnsmasqleases' => {
            'title' => "DHCP dnsmasq",
            'widget' => \&dnsmasqLeasesWidget,
            'order' => 5,
            'default' => 1
        }
    };
}




1;
